R/eml.R

```diff
@@ -1006,3 +1006,125 @@ reorder_pids <- function(pid_list, doc){
   ordered_pids <- pid_list[order(match(names(pid_list), entity_names))]
   return(ordered_pids)
 }
+
+#' Create an EML project section from a list of NSF award numbers
+#'
+#' This function takes a list of NSF award numbers and uses it to
+#' query the NSF API to get the award title, PIs, and coPIs. The
+#' return value is an EML project section. The function supports 1
+#' or more award numbers
+#'
+#' @param awards (list) A list of NSF award numbers as characters
+#' @param eml_version (char) EML version to use (2.1.1 or 2.2.0)
+#' @return project (emld) An EML project section
+#'
+#' @export
+#'
+#' @examples
+#' awards <- c("1203146", "1203473", "1603116")
+#'
+#' proj <- eml_nsf_to_project(awards, eml_version = "2.1.1")
+#'
+#' me <- list(individualName = list(givenName = "Jeanette", surName = "Clark"))
+#'
+#' doc <- list(packageId = "id", system = "system",
+#'            dataset = list(title = "A Mimimal Valid EML Dataset",
+#'                           creator = me,
+#'                           contact = me))
+#'
+#' doc$dataset$project <- proj
+#'
+#' EML::eml_validate(doc)
+#'
+eml_nsf_to_project <- function(awards, eml_version = "2.1"){
+
+  stopifnot(is.character(awards))
+  stopifnot(eml_version %in% c("2.1", "2.1.1", "2.2", "2.2.0"))
+
+  award_nums <- awards
+
+  result <- lapply(award_nums, function(x){
+    url <- paste0("https://api.nsf.gov/services/v1/awards.json?id=", x ,"&printFields=coPDPI,pdPIName,title")
+
+    t <- jsonlite::fromJSON(url)
+
+    if ("serviceNotification" %in% names(t$response)) {
+      warning(paste(t$response$serviceNotification$notificationType, "for award", x , "\n this award will not be included in the project section."), call. = FALSE)
+      t <- NULL
+    }
+    else if (length(t$response$award) == 0){
+      warning(paste("Empty result for award", x, "\n this award will not be included in the project section."), call. = FALSE)
+      t <- NULL
+    }
+    else t
+  })
+
+  i <- lapply(result, function(x) {!is.null(x)})
+  result <- result[unlist(i)]
+  award_nums <- award_nums[unlist(i)]
+
+  if (length(award_nums) == 0){
+    stop(call. = F,
+         "No valid award numbers were found.")
+  }
+
+  # create function to extract first name and middle initial (if present)
+  # as firstName, and whatever else exists as lastName
+  extract_name <- function(x){
```

Comment:
should this be defined outside the function body as a helper?

Reply:
hm. I don't think so. This is pretty specific to how the NSF API formats the names and I can't imagine it would be that useful anywhere else. Maybe it isn't good practice to define a function within another function though?

Reply:
That was my thought but it doesn't really matter.  No need to move it.